lib.js

```diff
@@ -0,0 +1,76 @@
+let { readdir, stat } = require("node:fs/promises");
+let path = require("node:path");
+/** @import { FileFinderOptions } from "./types.ts" */
+
+exports.FileFinder = class FileFinder {
+	/**
+	 * @param {string} root
+	 * @param {FileFinderOptions} options?
+	 */
+	constructor(root, { skipDotfiles, filter }) {
+		this._root = root;
+
+		/**
+		 * @param {string} filename
+		 * @return {boolean}
+		 */
+		this._filter = filename => {
+			if(skipDotfiles && isDotfile(filename)) {
+				return false;
+			}
+			return filter ? filter(filename) : true;
+		};
+	}
+
+	/**
+	 * a list of relative file paths within the respective directory
+	 *
+	 * @returns {Promise<string[]>}
+	 */
+	async all() {
+		let filenames = await tree(this._root);
+		return filenames.filter(this._filter);
+	}
+
+	/**
+	 * all file paths that match the filter function
+	 *
+	 * @param {string[]} filepaths
+	 * @returns {Promise<string[]>}
+	 */
+	async match(filepaths) {
+		return filepaths.map(filepath => path.relative(this._root, filepath)).
+			filter(filename => !filename.startsWith("..")).
+			filter(this._filter);
+	}
+};
+
+/**
+ * flat list of all files of a directory tree
+ *
+ * @param {string} filepath
+ * @param {string} referenceDir
+ * @returns {Promise<string[]>}
+ */
+async function tree(filepath, referenceDir = filepath) {
+	let stats = await stat(filepath);
+
+	if(!stats.isDirectory()) {
+		return [path.relative(referenceDir, filepath)];
+	}
+
+	let entries = await Promise.all((await readdir(filepath)).map(entry => {
+		return tree(path.join(filepath, entry), referenceDir);
+	}));
+	return entries.flat();
+}
+
+/**
+ * Does the filename start with a dot?
+ *
+ * @param {string} filename
+ * @returns {boolean}
+ */
+function isDotfile(filename) {
```

Comment:
Why not inline this? It's only ever used once.

Reply:
Tried it out, looks readable.
f4b8459bd15b8c51d12c0aec67270d02d11d84f9